DoIts without method header and reformatting

dionisiydk opened the pull request:
DoIts without method header and reformatting:
![simpleDoIt](https://user-images.githubusercontent.com/8149664/180092149-d8c28226-78a8-4ea9-b474-90bccc94be8a.png)

- RBParser is refactored to simplify method and doIts parsing/compilation
 - common parts are extracted to parseMethodBodyInto:
    - pragmas are supported in doIts (see #testPragmaInExpression):
      - it was some strange constraint

      - compilation flags can be now enabled in workspace
  - RBReturnNode is improved with better var name for position: #start instead of #return 
  - RBDoItMethodNode is introduced to represent method nodes in doIt ASTs
 - SHRBTextStyler>>isWorkspace uses isScripting which respects semantic scope to correctly highlight DoIts without headers

PR does not fix highlighting in the debugger and inspector for no-header DoIts. These fixes are in their repos:
- https://github.com/pharo-spec/NewTools/pull/400
- https://github.com/pharo-spec/Spec/pull/1296

MarcusDenker commented:
Only failing test is 

- SpOptionListPresenterTest>>testDoAcceptIsNotExecutedWhenCancelingDialog

Which is not related to the change

MarcusDenker commented:
Nice speedup for #evaluate:

```
[ Smalltalk compiler evaluate: '1+2' ] bench

before:  "'4431.741 per second'"
after:  "'28202.360 per second'"
```

dionisiydk committed:
DoIts without method header and reformatting:
        - RBParser is refactored to simplify method and doIts parsing/compilation
                - common parts are extracted to parseMethodBodyInto:
                - pragmas are supported in doIts (see #testPragmaInExpression):
                        - it was some strange constraint
                        - compilation flags can be now enabled in workspace
                - RBReturnNode is improved with better var name for position: #start instead of #return
               - RBDoItMethodNode is introduced to represent method nodes for doIts.
        - SHRBTextStyler>>isWorkspace uses isScripting which respects semantic scope to correctly highlight DoIts without headers

dionisiydk commented on src/AST-Core/ASTCacheMissStrategy.class.st:
```diff
@@ -16,5 +16,5 @@ Class {
 { #category : #accessing }
 ASTCacheMissStrategy >> getASTFor: aCompiledMethod [
-	^ aCompiledMethod parseTree doSemanticAnalysisIn: aCompiledMethod methodClass
+	^ aCompiledMethod parseTree
```
With little fix in #parseTree method (see below) the extra semantic analysis is not needed any more.

dionisiydk committed:
Class comment for RBDoItMethodNode

dionisiydk commented on src/DrTests-CommentsToTests-Tests/CommentsToTestsTest.class.st:
```diff
@@ -28,7 +28,7 @@ CommentsToTestsTest >> testCommentWithSyntaxError [
 	commentTestCase := CommentTestCase for: docComment.
 	
-	self should: [commentTestCase testIt] raise: SyntaxErrorNotification
+	self should: [commentTestCase testIt] raise: RuntimeSyntaxError
```
This difference is from the previous reformatting logic of ast methods. There were two stages:
1. initial parsing of source code in faulty mode
2. reformatting and parsing again with normal mode which causes syntax notification.

Now the second step is removed and therefore only runtime syntax error can be thrown.

MarcusDenker replied:
I always wanted to change the doc comments to store the AST, not the source, but that is for another PR of course

dionisiydk commented on src/OpalCompiler-Core/OpalCompiler.class.st:
```diff
@@ -259,17 +259,7 @@ OpalCompiler >> compile: textOrString [
 { #category : #private }
 OpalCompiler >> compileDoItFromAST [ 
```
This is the main change to remove the header and reformatting

MarcusDenker commented:
- ClassHierarchyPrinterTest>>testOnlyRBASTNodes needs to be update for the DoItMethodNode

dionisiydk committed:
Fix testOnlyRBASTNodes which hardcodes the actual RBNode hierarchy
